Sources/Overload/OvCore/src/OvCore/Rendering/ShadowRenderPass.cpp

```diff
@@ -0,0 +1,121 @@
+/**
+* @project: Overload
+* @author: Overload Tech.
+* @licence: MIT
+*/
+
+#include "OvCore/Rendering/ShadowRenderPass.h"
+
+#include <OvCore/ECS/Components/CMaterialRenderer.h>
+#include <OvCore/Rendering/EngineDrawableDescriptor.h>
+#include <OvCore/Rendering/EngineBufferRenderFeature.h>
+#include <OvRendering/Features/LightingRenderFeature.h>
+#include <OvCore/ResourceManagement/ShaderManager.h>
+#include <OvCore/Global/ServiceLocator.h>
+
+constexpr uint8_t kMaxShadowMaps = 1;
+
+OvCore::Rendering::ShadowRenderPass::ShadowRenderPass(OvRendering::Core::CompositeRenderer& p_renderer) :
+	OvRendering::Core::ARenderPass(p_renderer)
+{
+	const auto shadowShader = OVSERVICE(OvCore::ResourceManagement::ShaderManager).GetResource(":Shaders\\Shadow.ovfx");
+	OVASSERT(shadowShader, "Cannot find the shadow shader");
+
+	m_shadowMaterial.SetShader(shadowShader);
+	m_shadowMaterial.SetFrontfaceCulling(false);
+	m_shadowMaterial.SetBackfaceCulling(false);
+}
+
+void OvCore::Rendering::ShadowRenderPass::Draw(OvRendering::Data::PipelineState p_pso)
+{
+	using namespace OvCore::Rendering;
+
+	OVASSERT(m_renderer.HasDescriptor<SceneRenderer::SceneDescriptor>(), "Cannot find SceneDescriptor attached to this renderer");
+	OVASSERT(m_renderer.HasFeature<OvCore::Rendering::EngineBufferRenderFeature>(), "Cannot find EngineBufferRenderFeature attached to this renderer");
+	OVASSERT(m_renderer.HasDescriptor<OvRendering::Features::LightingRenderFeature::LightingDescriptor>(), "Cannot find LightingDescriptor attached to this renderer");
+
+	auto& engineBufferRenderFeature = m_renderer.GetFeature<OvCore::Rendering::EngineBufferRenderFeature>();
+	auto& lightingDescriptor = m_renderer.GetDescriptor<OvRendering::Features::LightingRenderFeature::LightingDescriptor>();
+
+	auto& sceneDescriptor = m_renderer.GetDescriptor<SceneRenderer::SceneDescriptor>();
+	auto& frameDescriptor = m_renderer.GetFrameDescriptor();
+	auto& scene = sceneDescriptor.scene;
+
+	auto pso = m_renderer.CreatePipelineState();
+
+	uint8_t lightIndex = 0;
+
+	for (auto lightReference : lightingDescriptor.lights)
+	{
+		auto& light = lightReference.get();
+
+		if (light.castShadows)
+		{
+			if (lightIndex < kMaxShadowMaps)
+			{
+				if (light.type == OvRendering::Settings::ELightType::DIRECTIONAL)
+				{
+					light.UpdateShadowData(frameDescriptor.camera.value());
+					const auto& lightSpaceMatrix = light.GetLightSpaceMatrix();
+					const auto& shadowBuffer = light.GetShadowBuffer();
+					m_shadowMaterial.Set("_LightSpaceMatrix", lightSpaceMatrix);
+					shadowBuffer.Bind();
+					m_renderer.SetViewport(0, 0, light.shadowMapResolution, light.shadowMapResolution);
+					m_renderer.Clear(true, true, true);
+					DrawOpaques(pso, scene);
+					shadowBuffer.Unbind();
+				}
+				else
+				{
+					// Other light types not supported!
```

Comment:
We could output an `OVLOG_WARNING("...");` to track-down if `castShadows` is set for unsupported lights. It could be seen as a "wrong setting" for those other light types

Reply:
The option for other lights to cast shadow isn't exposed, so unless a user write custom c++ code, this shouldn't happen. I agree that a warning wouldn't hurt tho

Reply:
Actually, if we were to add a warning here, it would log every frame, which could quickly become annoying x)

Reply:
The other option would be to warn once the invalid setting is set, but it isn't possible since it's stored in a struct, we can't provide any validation. So... I would say we leave it as-is 😅